fix: allow type imports in moduleResolution: node

JoshuaKGoldberg opened the pull request:
## PR Checklist

- [x] Addresses an existing open issue: fixes #7185
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Package types lookup in `"moduleResolution": "node"` adheres to Node 10 (`"moduleResolution": "node10"`), which doesn't support `"exports"` > `"."` > `"types"` field in the `package.json`. https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/a8dece2e3aab25f9ddbef1ac2f1c22ae8d27bc6f/docs/problems/NoResolution.md#true-positive-node-10-doesnt-support-packagejson-exports recommends using a newer `moduleResolution` in your TSConfig. To help out users still using Node 10 module resolution, this adds explicit `"types": "./(dist/)index.d.ts"` entries to the `package.json`s.

Note that you can avoid the need for this code change by setting a new module resolution strategy, such as `"moduleResolution": "node16"`, in your TSConfig:

```diff
{
  // node-lts-strictest is deprecated in favor of using an array to combine node-lts and strictest but doing so
  // currently leads to problems with the eslint-plugin-import package, see
  // https://github.com/import-js/eslint-plugin-import/issues/2751
  "extends": "@tsconfig/node-lts-strictest/tsconfig.json",
  "compilerOptions": {
+    "moduleResolution": "Node16",
  },
}

```

bradzacher reviewed (commented):
This was an intentional removal though because we don't support node10 - we're now at a minimum of node16 - so supporting types for node10 resolution is wrong and allows people to do broken things (eg TS will happily let you do deep imports, but then that code will crash at runtime).

JoshuaKGoldberg commented:
Ah, you're right - I forgot about this convoluted chain.

tl;dr that I'll copy & paste around:

* We dropped support for Node <=14
* As a part of ☝️ we removed the entries in our `package.json`s that helped old Node versions resolve package entries ("module resolution")
* TypeScript respects Node module resolution algorithms, which means we essentially dropped _types_ module resolution support for older versions of Node
* TSConfig `"moduleResolution": "node"` is actually an alias of `"node10"`, which is EOL - you should switch to `"node16"` or `"nodeNext"` or whatever is most accurate for your project per https://www.typescriptlang.org/tsconfig#moduleResolution

Working as intended.
